### PR TITLE
add derivation for the virtualbox oracle extension pack

### PR DESCRIPTION
**Pull request by cdepillabout**
###### Motivation for this change

This PR implements the first stage of the plan laid out by @flokli for managing the VirtualBox Oracle Extension Pack as a separate derivation:

https://github.com/NixOS/nixpkgs/issues/34796#issuecomment-365070156

Currently, the VirtualBox Oracle Extension Pack needs to be downloaded directly by the end-user with the command `nix-prefetch-url`.  The `virtualbox` derivation takes a boolean `enableExtensionPack` flag that the user needs to set.

This PR makes the following changes/additions:

1.  Add the Virtualbox PUEL to the licenses file.  It has been added with `free = false`.  This means that only uses that have `config.allowUnfree = true` will be able to use it.
1.  Split the VirtualBox Oracle Extension Pack into a separate derivation.  It is now available in the `all-packages.nix` file as an attribute called `virtualboxExtpack`.
1.  Change the main `virtualbox` derivation so that instead of taking in a boolean flag called `enableExtensionPack`, it instead just takes a derivation it calls `extensionPack`.  When specified, this will normally just be the top-level `virtualboxExtpack`.
1.  Change the NixOS `virtualbox-host` module to enable the extension pack when `virtualisation.virtualbox.host.enableExtensionPack` is set to `true`.  This of course won't work unless `allowUnfree` is set to `true`.

Note that as explained in #34796, the end-user will still have to compile VirtualBox on their own machine.  However, they will NOT have to download the Oracle Extension Pack by hand.  I see this as an improvement.

This PR lays the ground-work for future improvements around VirtualBox and the Oracle Extension Pack.

Here is a NixOps file that can be used to test this change:

```nix
{
  network.description = "NixOS virtualbox test";

  # This is the configuration for a specific machine.  It follows the NixOS
  # configuration specification, which can be found below:
  # https://nixos.org/nixos/manual/index.html#ch-configuration
  vbox-test-server =
    { config, pkgs, ...}:
    {
      imports = [ ./vbox-test.nix ];

      deployment.targetEnv = "virtualbox";
      # deployment.virtualbox.headless = true;
      deployment.virtualbox.memorySize = 2024;
      deployment.virtualbox.vcpu = 1;
    };
}
```

And the corresponding `vbox-test.nix` file:

```nix
{ config, lib, pkgs, ... }:
{
  nixpkgs.config.allowUnfree = true;
  virtualisation.virtualbox.host.enable = true;
  virtualisation.virtualbox.host.enableExtensionPack = true;

  environment.systemPackages = with pkgs; [
    bash
    gitAndTools.gitFull
    firefoxWrapper
    xterm
  ];

  services.xserver.enable = true;
  services.xserver.layout = "us";
  services.xserver.desktopManager = {
    xfce.enable = true;
    default = "xfce";
  };
}
```

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by cdepillabout** [comment]
I should also ping @svanderburg since he is a maintainer for VirtualBox.

File: pkgs/top-level/all-packages.nix
Comment (flokli) [inline pkgs/top-level/all-packages.nix]: why do we want to have the `lowPrio` here? Just to make it more deterministic if somebody installs both `virtualbox` and `virtualboxWithExtpack` which one will win?

**Review by flokli — COMMENTED** [review]
Thanks for this PR!

What's your opinion on adding the `vbox-test.nix` file mentioned in the PR description too, so somebody doing changes can locally run these tests (or a variant of the already existing virtualbox test, but calling it with `virtualboxWithExtpack`)

File: pkgs/top-level/all-packages.nix
Comment (cdepillabout) [reply pkgs/top-level/all-packages.nix]: @flokli Thanks for reviewing this!

To be honest, I'm actually not sure what role `lowPrio` is playing.

You can't tell because Github cuts off the `all-packages.nix` file, but there are actually a couple different versions of the virtualbox derivation:

```nix
  virtualbox = callPackage ../applications/virtualization/virtualbox {
    stdenv = stdenv_32bit;
    inherit (gnome2) libIDL;
    pulseSupport = config.pulseaudio or true;
  };

  virtualboxHardened = lowPrio (virtualbox.override {
    enableHardening = true;
  });

  virtualboxHeadless = lowPrio (virtualbox.override {
    enableHardening = true;
    headless = true;
  });
```

I setup `virtualboxWithExtpack` to use `lowPrio` because that's what `virtualboxHardened` and `virtualboxHeadless` derivations were doing.

I'm not sure if it is "correct" to use `lowPrio` here or not.

If you think that `lowPrio` shouldn't be used here, definitely let me know!

**Comment by cdepillabout** [comment]
> What's your opinion on adding the vbox-test.nix file mentioned in the PR description too, so somebody doing changes can locally run these tests (or a variant of the already existing virtualbox test, but calling it with virtualboxWithExtpack)

I'm relatively new with NixOS.  This is the first non-trivial PR I've sent recently.  I've seen that there are `pkgs/test/` and `nixos/tests/` directories, but I've never actually looked inside them!  I don't know enough about how the tests work to know whether or not my `vbox-test.nix` file should be added.

It looks like there is a section on [testing](https://nixos.org/nixos/manual/index.html#sec-nixos-tests) in the NixOS manual, so I'll take a look at that and then take a look at the existing VirtualBox tests.

File: nixos/modules/virtualisation/virtualbox-host.nix
Comment (zimbatm) [inline nixos/modules/virtualisation/virtualbox-host.nix]: How is the license being accepted? It seems like it's possible to  set `config.allowUnfree = true` without having read the license. It's a blanked agreement to all unfree licenses. Usually EULA require an explicit approval (which is questionable, but that's what is happening).

Comment (zimbatm) [reply nixos/modules/virtualisation/virtualbox-host.nix]: One idea would be that the user has to provide the sha256 of the license as a "proof" that he read and accepted it

Comment (cdepillabout) [reply nixos/modules/virtualisation/virtualbox-host.nix]: Hi @zimbatm, thanks for taking a look at this PR.

> How is the license being accepted?

Originally (the current state of nixpkgs), if you enable `virtualization.virtualbox.host.enableExtensionPack`, `nix-build` will fail and you are presented with a message telling you to read the PUEL and asking you to download the download the Oracle Extension Pack directly by using `nix-prefetch-url`.

This PR changes it so that the license is accepted and the Extension Pack will be downloaded during `nix-build` if you set both `config.allowUnfree = true` and `virtualization.virtualbox.host.enableExtensionPack = true`.

> One idea would be that the user has to provide the sha256 of the license as a "proof" that he read and accepted it

As I said in the original issue, I don't know the legal ramifications of setting it up to work one way vs another way.

However, I'm not sure I understand the reasoning for handling the PUEL as a somewhat special case, and requiring the end user to do something like provide a `sha256` of the license.

I took a look at another random non-free license, and it seems to have similar wording to the PUEL.  Here is the license for the Elastic License, which is used in the non-free version of kibana, logstash, etc:

https://github.com/elastic/elasticsearch/blob/master/licenses/ELASTIC-LICENSE.txt

Here is the PUEL for comparison:

https://www.virtualbox.org/wiki/VirtualBox_PUEL

Here is the page that allows you download the Extension Pack and links to the above license:

https://github.com/elastic/elasticsearch/blob/master/licenses/ELASTIC-LICENSE.txt

----------------------------------------------------------------------------------------

With that said, if everyone agrees that the PUEL is special in this regard, and we should do something like require the user to provide a sha256 of the license text, then I would be happy to change this PR to do that!

My main goal is to get rid of the manual steps in installing Virtualbox with the Extension Pack.

**Comment by cdepillabout** [comment]
@flokli 

In commit 2ae9907 I've added a test to make sure that the VirtualBox Extension Pack is successfully installed and can be used.

This test can be run like the following:

```sh
$ cd nixpkgs/nixos/tests/
$ nix-build virtualbox.nix -A enable-extension-pack
```

You can run all the virtualbox tests like the following:

```sh
$ nix-build virtualbox.nix
```

File: nixos/modules/virtualisation/virtualbox-host.nix
Comment (zimbatm) [reply nixos/modules/virtualisation/virtualbox-host.nix]: Okay so this is complementary. `pkgs/stdenv/generic/check-meta.nix` there is a `allowUnfreePredicate` that would allow to build a list of authorized licenses.

I think the issue is around the notion of explicit consent, a user should not be able to use software on which he hasn't agreed on the license. `allowUnfree = true` is quite a low bar to say "Yes I have read and understand all the unfree license and I agree with all of them".

But like you say the mechanism is like that currently, it's independent and shouldn't hold back this PR.

**Comment by cdepillabout** [comment]
This PR had a few conflicts with master.  They have been fixed in commit e04e92d.

Unless anyone else notices anything that can bemajorly improved, I think this PR should be ready to be merged as-is.

It would be great to get it merged in time for 18.09.

**Review by zimbatm — APPROVED** [review]
LGTM!

**Comment by aszlig** [comment]
@cdepillabout, @zimbatm: From the [PUEL](https://www.virtualbox.org/wiki/VirtualBox_PUEL):

> (2) You may not do any of the following: (a) modify any part of the Product, except to the extent allowed in the documentation accompanying the Product; (b) rent, lease, lend, **re-distribute**, or encumber the Product; (c) remove or alter any proprietary legends or notices contained in the Product; or (d) decompile, or reverse engineer the Product (except to the extent permitted by applicable law).

IANAL, but I think the PUEL is pretty clear about that, so I think we should not re-distribute the extension pack. Unfortunately since this adds a test with the extension pack activated, we now do exactly that:

```sh-session
$ nix path-info --store https://cache.nixos.org/ --json /nix/store/6f470485gpl6ri8kv7dzii7nyngxk40d-Oracle_VM_VirtualBox_Extension_Pack-5.2.14-123301.vbox-extpack | jq
[
  {
    "path": "/nix/store/6f470485gpl6ri8kv7dzii7nyngxk40d-Oracle_VM_VirtualBox_Extension_Pack-5.2.14-123301.vbox-extpack",
    "narHash": "sha256:03js53j30ddpfy1l7cg0qyr0iwnmmxv1qcjdnsrgyrf25javmxfw",
    "narSize": 19589328,
    "references": [],
    "deriver": "/nix/store/3l4m8vjmf6rqjvzknxvn4lradjjlfdsm-Oracle_VM_VirtualBox_Extension_Pack-5.2.14-123301.vbox-extpack.drv",
    "signatures": [
      "cache.nixos.org-1:JW0RhCmX5Jn8/Zq5BAByVjXqMnyADKSLuPyeO9wPjtDvDIBIGlguijePcmDJzqcFPvBWqFIPxa2RNcB53nW5Bw=="
    ],
    "url": "nar/1px2nyzd47x3hwc6z43ir3xvv2xhjc0saiw2ziyznbprs63zj7zs.nar.xz",
    "downloadHash": "sha256:1px2nyzd47x3hwc6z43ir3xvv2xhjc0saiw2ziyznbprs63zj7zs",
    "downloadSize": 19240484
  }
]
```

**Comment by zimbatm** [comment]
good point @aszlig. @cdepillabout can you make another PR that disables the test if allowUnfree is turned off?

**Comment by cdepillabout** [comment]
@flokli @zimbatm Sounds good, in the next couple days I'll send a PR modifying the virtualbox tests to work like the ELK tests linked above.

**Comment by grahamc** [comment]
I wonder why the NixOS tests will build tests with nonfree dependencies -- that seems like the real bug here.

**Comment by aszlig** [comment]
@grahamc: It won't, that's why this PR added [nixpkgs.config.allowUnfree](https://github.com/cdepillabout/nixpkgs/blob/e04e92d38b944dd5729ed023f9f5e131acf0e95a/nixos/tests/virtualbox.nix#L350).

**Comment by grahamc** [comment]
I see. We should probably not do that?

**Comment by cdepillabout** [comment]
> I see. We should probably not do that?

@grahamc This PR is adding a derivation that will only compile if `allowUnfree` is set to true.  The test is testing this derivation, so `allowUnfree` needs to be set to true to actually build it.

As I understand it, it seems like there are two problems:

1. The tests probably shouldn't set `allowUnfree` to true.  If they DO need to do this, it should probably be behind a configuration option (similar to how the ELK tests linked above work).  I am planning on sending a PR to fix this for the virtualbox tests.
2. Even if the tests do set `allowUnfree` to true, cache.nixos.org should either refuse to run them, or not distribute the unfree derivations.  It sounds like you are suggesting that this is also a bug that needs to be fixed.

**Pull request merged**
